src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets

```diff
@@ -128,4 +128,25 @@
       Condition="'$(_symbolIsFile)' == 'false' and Exists('$(_symbolSourcePath)')" />
   </Target>
 
+  <Target Name="_WriteLinkingInstructionsAsMarkdown" AfterTargets="_CopyAotSymbols"
+          Condition="'$(NativeLib)' == 'Static' and '$(SkipLinkingInstructions)' != 'true'">
+
+    <ItemGroup>
+      <FileAsset Include="$(PublishDir)\$(TargetName)$(NativeBinaryExt)" />
+      <FileAsset Include="@(CustomLinkerArg)" Condition="Exists('%(CustomLinkerArg.Identity)') and '%(CustomLinkerArg.ExcludeFromStaticInstructions)' != 'true'" />
+      <LinkOption Include="@(CustomLinkerArg)" Condition="!Exists('%(CustomLinkerArg.Identity)') and '%(CustomLinkerArg.ExcludeFromStaticInstructions)' != 'true'" />
```

Comment:
I wonder if it would make sense to lean into the linker argument separation we started for Xamarin/MAUI in https://github.com/dotnet/runtime/pull/88294#issuecomment-1663268829 and https://github.com/dotnet/runtime/pull/89916. The separation should already be there on mac and we might need to put a couple things into different ItemGroup on Linux.

Basically instead of doing a heuristic filtering on CustomLinkerArg, we'd compose these lists from the ~5 properties and ItemGroups.

This heuristic would also work, but we'll probably end up with many ExcludeFromStaticInstructions exclusions (Linux has more switches that are in the unncessary/potentially problematic bucket - and even on mac, we'd probably want to exclude rpath, for example).

Reply:
I have tested it on both mac and linux and tried the resultant lists usage with `eval cc myglue.o -o myexe $(cat dist/sources.txt) $(cat dist/linkoptions.txt)` using [this program](https://github.com/dotnet/runtime/issues/70277#issue-1261727711). It links with clang on mac and linux and gcc on linux drivers and bfd linker on linux (lld runs into https://github.com/dotnet/runtime/issues/70277#issue-1261727711, I'll investigate it separately). Both on linux and mac I'm getting:

```
Hello from C# Add!
Hello from C callback!
sum is: 3
```

These options are ok to make linked app just work. If they want to customize it further, like generating a static binary out of the NativeLib=Static, they will need to set StaticICULinking, StaticOpenSslLinking etc. and pass -static to link options. While we can't predict and automate for all kinds of use-cases, just a head-start, we can definitely minimize this set to bare minimum and filter out the rest to let user make their own choices.

The problem with separate lists is that `LinkerArg` is something used in the wild. Users have their own options set.  That's why I thought of using this inference. If we are going to split them and document the "sources" vs. "options" msbuild items, that would be great, but as it stands, I couldn't think of better way.

Reply:
> lld runs into [#70277 (comment)](https://github.com/dotnet/runtime/issues/70277#issue-1261727711), I'll investigate it separately

I take it back. To use lld, we actually need to publish with `-p:LinkerFlavor=lld` (so we get `-Wl,-T,"obj/Release/net9.0/linux-musl-arm64/native/sections.ld"` in linkoptions.txt).

Note that if user published with LinkerFlavor=lld and then decided to use bfd to link with their own code, it gives syntax error (all linkers have seem to invented their own layout script format and there is no consensus). So they will need to just drop the `-Wl,-T` line.

Reply:
> I wonder if it would make sense to lean into the linker argument separation we started for Xamarin/MAUI 

I think this is a good point. Do we believe that Xamarin/MAUI can switch over to this scheme? They are the real-world use for static linking. If they are not, it suggests that this is not the best design. It would be best to avoid creating and maintaining multiple recipes for static linking.

Reply:
@jkotas do we have a publicaly visible link where they are using `NativeLib=static`? Maybe they are wrapping it under mono library-mode and bypassing the BuildIntegration?

Also, for classification, we would need to consider `<LinkerArg` in user's csproj where this approach seems to be working.

Reply:
https://github.com/xamarin/xamarin-macios/blob/002eeddaf42a4b8c56631cc98d5f072a89926083/dotnet/targets/Xamarin.Shared.Sdk.targets#L1364-L1368 is what the Xamarin static linking integration looks like. I do not think they actually need to set `NativeLib=static` to make it work.

Reply:
Not sure what you mean, can you explain? These instructions are only enabled for NativeLib=static, so if they are not using this mode, none of this impact/benefit them.

Reply:
They are not using the `NativeLib=static` mode, but they are doing what this mode is trying to achieve: take .o file produced by naot build and link it together with naot runtime and a bunch of non-.NET stuff into a single binary.

We have two recipes to how to do it: Xamarin iOS one (we do not have written down in a simple form - something to fix),  and the one being designed in this PR. Do we need both? If we need two, what's our guideline for choosing which one of these recipes to use?

Reply:
This mode (NativeLib=static) produces shared lib (`.a` archive file) in publish directory. The helper lists generated by this PR will help users to link that `.a` file with their code.

Xamarin's infra look to be a bit advanced or non-traditional that it takes the `.o` file (which PublishAot=true generates under obj/ dir) and they put together other runtime/libraries pieces by themselves instead of using BuildIntegration.

I agree if Xamarin doesn't have any special requirement and can use .a out of standard PublishAot flow, we should aim to utilize these helpers in their infra. However, I am not sure if it is feasible for their use-case to align with PublishAot workflow.

Reply:
> This mode (NativeLib=static) produces shared lib (.a archive file) in publish directory.

Is there a point in the extra step that produces the .a file? Should this flow take the .o file like the Xamarin flow instead?

Reply:
`.a` static library, reducing file size by 37% compared to `.o` with helloworld, adheres to standard practices, optimizing distribution and maintaining efficient linking processes. I think Xamarin should try to adapt to this mechanism from nativeaot/BuildIntegration, instead of NativeAOT publish process molded to Xamarin's internals. We have several flags controlled by NativeLib==/!=static which Xamarin targets skip over and then they uses their own `LinkNativeCode` MSBuild task to perform the linking.

Reply:
Why is static library reducing size compared to .o? Everything that we emit into .o is expected to be necessary.

Reply:
Archive `.a` is the static library format on Unix-like systems and a well-established standard for distributing compiled code.  Whereas, `.o` is the intermediate form, object file. The linker knows how to read indexed symbol table out of `.a` for efficient linkage. With NativeLib=static, user expects and gets a static library.

Reply:
I do not think that anybody should be distributing these static libraries. It is too fragile system to be distributed and it does not behave like a regular static library (e.g. you cannot link two these to a project - they would collide).

This is only meant to be used locally and tightly integrated into some larger project build.

Reply:
Maybe we should stop calling this "static libraries" to avoid implying that the output behaves like a regular static library.

Reply:
> e.g. you cannot link two these to a project - they would collide

That problem also exists with object files and the solution is the same; don't use multiple of them and put all code in a single compilation unit or pass `-Wl,--allow-multiple-definition` to the linker.